Interface guidelines for forms (#261)
primer/design: content/ui-patterns/forms.mdx

Fields must be labeled so users understand what the field is used for.

Labels should be descriptive and as short as possible: ideally, it will always fit on one line of text. Labels should be written in sentence-case. For example: "Repository name".

If you're having trouble keeping label text short, consider using a [caption](#caption) to provide more context.

In rare cases, a label may be hidden when the user has enough context to understand what the field does. When a label is visually hidden, label text must still be provided for assistive technology such as screen readers. For example: a search field with an icon that is placed in the spot a user would expect to find a search field may have a visually hidden label that reads "Search"

### Required field indicator
When a field is required to have a value, it should be visibly marked as required. For more information, see the [required fields](#required-fields) section.

### Input (required)
The input is what the user interacts with to set the value of the field.

For more specific information about each input, visit the respective component documentation page. For example: [CheckboxInputField](), [TextInputField]()

> ### Placeholder text

Reviewer: @mperrotti I mentioned this in the TextInput component review, [but I proposed we completely do away with Placeholder text](https://github.com/github/primer/issues/538#:~:text=a%20loading%20state-,Placeholder%20text,-%F0%9F%94%97). [@jscholes had a similar suggestion](https://github.com/github/accessibility/issues/892#issuecomment-1077764961) as well. 
- https://github.com/github/primer/issues/538#:~:text=a%20loading%20state-,Placeholder%20text,-%F0%9F%94%97

Author: Thanks for the info. I'm going to remove the "Do / Don't" example so we're not explicitly encouraging the use of `placeholder`.

---

> title: Forms

Author: I always forget these. I **will** remember the next one!!

---

import {Box, Text} from '@primer/components'

<Box sx={{fontSize: 3}} class="lead" as="p">
    Forms are used to complete tasks that require data input from the user. For example: creating a new repo, configuring settings, and logging in.
</Box>

<Box sx={{fontSize: 3}} class="lead" as="p">
    Primer's form design guidelines aim to minimize the effort and cognitive load required to complete a task that involves a form.
</Box>

## Form control anatomy

![diagrams labeling the anatomy of a text field and a checkbox field](https://user-images.githubusercontent.com/2313998/164065184-e3c615a3-fb68-4878-b7f8-9a57efb748cb.png )

### Label (required)

> Labels should be descriptive, and as short as possible: ideally no more than 3 words. Labels should be written in sentence case. For example: "Repository name".

Reviewer 2: > Labels should be descriptive, and as short as possible

"As short as possible" feels like we prioritize sizing over semantics or meaning.

Reviewer 2: @mperrotti have you thought about this?

How about this?

```suggestion
Labels should be descriptive and concise: aim for no more than 3 words. Write labels in sentence case. Examples: "Repository name", "Payment method".
```

If you're having trouble keeping label text short, consider using a [caption](#caption) to provide more context.

> In rare cases, a label may be hidden when the user has enough context to understand what the input does. When a label is visually hidden, label text must still be provided for assistive technology such as screen readers. For example: a search field with an icon that is placed in the spot a user would expect to find a search field may have a visually hidden label that says "Search".

Reviewer 2: > In rare cases, a label may be hidden

> For example: a search field with an icon that is placed in the spot a user would expect to find a search field may have a visually hidden label that says "Search".

Is a search field part of a forms pattern? Search fields appear throughout dotcom, and they're not rare =).

Author: > Is a search field part of a forms pattern?

It _could_ be, but usually it isn't. I was trying to think of the most common and obvious example.

Do you (or any other reviewers) have any better examples?

Reviewer 2: @mperrotti I'm not sure. It's confusing to see an example about a search field while this page is specifically about [long-form] forms pattern.

Author: Good point about these guidelines being for long-form forms. I think it might be best to exclude this paragraph. Visually hidden labels can be addressed in form **component** guidelines.

> Placeholder text is never an acceptable substitute for a form control label because:

Reviewer 3: ```suggestion
Placeholder text is never an acceptable substitute for a label because:
```

- The placeholder text disappears as soon as the input has a value

> - Placeholder text colors are typically too light to meet the minimum color contrast ratio required to be accessible

Reviewer 3: ```suggestion
- Placeholder text colors are typically too light to meet the minimum color contrast ratio required for accessibility
```

- Screen readers do not read placeholder text

<DoDontContainer>
  <Do>
    <img role="presentation" src="https://user-images.githubusercontent.com/2313998/164065187-d300024d-ce39-4f5e-82ad-9e74c4b6d28a.png" />
    <Caption>Use a placeholder to hint at what kind of information is expected</Caption>
  </Do>
  <Dont>
    <img role="presentation" src="https://user-images.githubusercontent.com/2313998/164065188-35462873-3706-4063-b7b9-9d0993665307.png" />
    <Caption>Don’t use a placeholder as a substitute for a label</Caption>
  </Dont>
</DoDontContainer>

### Required field indicator

> When a field is required to have a value, it should be visibly marked as required. An individual checkbox or radio cannot be marked as required.

Reviewer 3: ```suggestion
When a field is required to have a value, it should be visibly marked as required. An individual checkbox or radio button cannot be marked as required.
```

For more information, see the [required fields](#required-fields) section.

### Input (required)

The input is what the user interacts with to set the value of the input. You may pre-fill inputs with smart default values, but be careful about making too many assumptions about what a user wants or needs.

> The input must have a minimum contrast ratio of 3:1 to pass WCAG AA-compliance. This is defined in [WCAG success criterion 1.4.11](https://www.w3.org/WAI/WCAG21/Understanding/non-text-contrast.html).

Reviewer 3: ```suggestion
The input must have a minimum contrast ratio of 3:1 to pass WCAG AA-compliance. Contrast requirements are defined in [WCAG success criterion 1.4.11](https://www.w3.org/WAI/WCAG21/Understanding/non-text-contrast.html).
```

Reviewer 2: Why is this guideline here? Shouldn't the input component provide the correct colors by default?

Author: @vdepizzol - I specifically called this out for when/if we decide to redesign our inputs.

This paragraph may need to be rewritten anyway. I looked at dotcom with all of the inputs having a 3:1 contrast, and it looked way too heavy. We may end up leaning on high-contrast color modes for those inputs.

dotcom test branch: [`mp/control-min-contrast`](https://github.com/github/github/pull/new/mp/control-min-contrast)

Author: I'm removing this until we make a decision about that how that WCAG guideline works with our color modes.

### Caption

A caption may be used to provide additional context about the field to help users fill in the correct data or explain how the data will be used. Caption text should be as short as possible.

Caption text may be displayed with a validation message, or it may be hidden if it is redundant information.

Caption text may be used to augment the [label](#label), but should not be redundant with the label or any other parts of the form control. If the caption feels redundant, try removing it. 

<DoDontContainer>
  <Do>

>     <img role="presentation" src="https://user-images.githubusercontent.com/2313998/164065190-5ef9d8db-c9ca-4aec-a2ea-321a81cc080e.png" />

Reviewer 2: The caption example feels off here:

> "Set automatically" will determine the closest available region based on your location (IP address)

Can we include something that fits closer with our [Content guidelines](https://primer.style/design/foundations/content)?

    <Caption>Use caption text that shows new information and provides helpful context</Caption>
  </Do>
  <Dont>
    <img role="presentation" src="https://user-images.githubusercontent.com/2313998/164065192-f1fd1c28-2031-4bb1-836b-607d14a3d4f4.png" />
    <Caption>Don’t use caption text that is redundant</Caption>
  </Dont>
</DoDontContainer>

> ### Validation message

Reviewer 2: The examples using checkbox groups have some weird application of bold. Do we really want to use bold labels for every checkbox or radio button item? Are the validation messages also meant to be bold? 🤔

Author: These are all meant to be bold. That's how the components are currently designed.

I'd be happy to change the checkbox and radio labels from bold to regular weight in separate PCSS and PRC PRs. What do you think @vdepizzol?

Author: The semibold error messages are kind of new: https://github.com/primer/react/pull/2013

@colebemis and I think it helps the text feel more balanced with the leading icon, but I'm open to changing it back to regular weight it if it's causing an issue.

Other reviewers - please chime in if you agree with @langermank that validation text should go back to regular weight.

Reviewer 2: @mperrotti I'd keep it at a regular weight. Validation messages can appear multiple times to the user in the same form. The bolded font feels too heavy.

Author: Ok, I'll update the example images and submit a PR to PRC to change the weight back to normal.

A single validation message may be displayed to provide helpful information for a user to complete their task. For example: explain an invalid value and submit the form.

> An individual checkbox or radio should not have its own validation.

Author: We don't visually mark an individual checkbox or radio as valid either. How about something like this?

```suggestion
An individual checkbox or radio should not have its own validation message or style.
```

<DoDontContainer>
  <Do>

>     <img role="presentation" src="https://user-images.githubusercontent.com/2313998/164065196-b0146ade-eccb-4a71-b1b6-f4a82b96c115.png" />

Reviewer 4: Are we supporting danger variants for checkboxes now?

Author: No

    <Caption>Show validation for the group of inputs</Caption>
  </Do>
  <Dont>
    <img role="presentation" src="https://user-images.githubusercontent.com/2313998/164065199-6cfa4893-55cb-4f81-a7d6-4b6ff66339eb.png" />
    <Caption>Don’t show validation for each input</Caption>
  </Dont>
</DoDontContainer>

Information from the caption should not be repeated in the error message. Show the validation message and remove the caption. A validation message makes the field easier to spot when a user is scanning for invalid fields.

<DoDontContainer>
  <Do>
    <img role="presentation" src="https://user-images.githubusercontent.com/2313998/164065196-b0146ade-eccb-4a71-b1b6-f4a82b96c115.png" />
    <Caption>Hide a caption that is redundant with the validation message</Caption>
  </Do>
  <Dont>
    <img role="presentation" src="https://user-images.githubusercontent.com/2313998/164065194-8f123836-5cda-4eea-8f0f-f1c016aa197c.png" />

>     <Caption>Don’t show a repetitive validation message</Caption>

Author: The caption should never change. It should either be shown after the validation message, or hidden.

In this case, the caption sounds repetitive after the validation message. I think this example is a little confusing, so I'll come up with something clearer.

  </Dont>
</DoDontContainer>

For more information about form validation, see the [validation guidelines](#invalid-input).

> ## Input methods

Reviewer 2: Shouldn't this section be structured as a list of components for easier parsing, instead of prose with static examples?

[Example](https://primer-6d38263007-26441320.drafts.github.io/components/action-list#application):

> ![Screenshot of Application section in Action list interface guidelines](https://user-images.githubusercontent.com/293280/166301732-1a450074-3321-4398-8aa5-3d9604d14b8a.png)

Author: I really like this. Thanks for the example!

### Open-ended text

![Text input, open-ended autocomplete text input, plain text input, textarea](https://user-images.githubusercontent.com/2313998/164782311-93a743fb-7757-4f29-bc74-522408a01b5c.png)

Use an open-ended text field when the field does not have a list of possible values. If it's possible to suggest values, use an [Autocomplete](#autocomplete) that allows users to pick a value or enter their own.

### A set of selectable options

![Action menu, autocomplete input with limited options, checkbox group, radio group, select](https://user-images.githubusercontent.com/2313998/164782313-851a5a90-8c50-409f-b03d-4200928e0a67.png)

Show a set of selectable options when there is a finite number of possible values.

## Structure

Forms should have a structure that makes it easy for users to scan. Form controls that are stacked vertically and left-aligned are easier for sighted users to scan visually.

<DoDontContainer>
  <Do>

>     <img role="presentation" src="https://user-images.githubusercontent.com/2313998/164065202-5c3aba0b-5cef-4ef4-9339-7192a3ebf23a.png" />

Reviewer 2: > <img src="https://user-images.githubusercontent.com/293280/166302029-b8201bb2-e4f6-4222-8436-e71bf655c866.png" width="400" alt="" />

The radio button and checkbox groups seem to have inconsistent headings and unnecessary bold text. Could we revisit these design decisions?

Author: Yes, let's revisit. I think the checkbox and radio labels should be the same weight as body text.

@vdepizzol - could you open an issue or discussion?

    <Caption>Default to vertically stacked form controls</Caption>
  </Do>
  <Dont>
    <img role="presentation" src="https://user-images.githubusercontent.com/2313998/164065205-4000430d-2216-4954-a367-86fbd0254cc3.png" />
    <Caption>Don’t lay out forms into columns just to reduce the vertical space used by the form</Caption>
  </Dont>

> </DoDontContainer>

Reviewer 4: @mperrotti is the text formatting for the radio selects and the label above the radio select group a new pattern? It feels like the bolding is switched and the group labels (e.g. `Preferred Primer component interface`) should be bold and the radio select options are normal weight.

Author: What's rendered here is what's rendered by our components. There has been some discussion below about changing the checkbox and radio labels to the normal weight.

### Order of form controls

Form fields should be in a predictable order that flows intuitively.

To achieve an intuitive flow:
- Order fields by their relative level of importance
- Keep related fields near each other
- When possible, keep inputs that require keyboard input near each other so that users who use a mouse don't have to switch back and forth between clicking and typing

### Grouping form controls

<Box mb={3} display="flex" alignItems="flex-start" flexDirection={["column", "column", "column", "column", "row"]} sx={{gap: 3}}>
    <img
        width="456"

>         alt="visually grouped of related inputs"

Reviewer 2: > ![image](https://user-images.githubusercontent.com/293280/166302520-ea11f1f8-a2d6-42ca-820d-f64faa4c0c4b.png)

I'm not sure I understand what this image is telling here —
* some headings are inside groups, some are not.
* Is "grouping" specifically referring to horizontal placement?
* Does Group 1.1 inside Group 1 specific which width it should follow?

Author: > some headings are inside groups, some are not.

I think you're referring to the labels in the first three form controls as "headings". None of these are headings, they are input labels.

---

> Is "grouping" specifically referring to horizontal placement?

Horizontal placement is just one of the ways you can visually group two or more inputs.

---

> Does Group 1.1 inside Group 1 specific which width it should follow?

I'm not sure I understand this question.

Author: I updated the image, but it's still a little "squished" because of the annotations.

        src="https://user-images.githubusercontent.com/2313998/164065207-16e51727-55b2-48bc-930c-ad9f0890ae10.png"
    />

>     <Box as="p" mt="0">When there is a collection of closely related fields, they should be labeled and visually grouped together. Effective techniques to visually group a collection of form controls include proximity and whitespace.</Box>

Reviewer 2: > Effective techniques to visually group a collection of form controls include proximity and whitespace.

Could we be more concise here? :))

In the image, there's a spacing of 8px between elements. I'd make sure we use 16px-equivalent spacing between controls in these scenarios.

Author: +1 for being more concise. Writing is hard 🙈

> In the image, there's a spacing of 8px between elements. I'd make sure we use 16px-equivalent spacing between controls in these scenarios.

So if 16px (`space.3`) is the spacing between closely related inputs, would we use 24px (`space.4`) for the default space between inputs? That seems a little big to me 🤔

</Box>

### Progressively disclosed form controls

<Box mb={3} display="flex" alignItems="flex-start" flexDirection={["column", "column", "column", "column", "row"]} sx={{gap: 3}}>

>     <img

Reviewer 2: > <img src="https://user-images.githubusercontent.com/293280/166303136-cc5d6cdb-494d-4836-8137-7c0930d36e71.png" alt="" width="400" />

Is this a realistic width we should recommend for controls? Everything looks so squished horizontally. Also, the spacing between expiration and CVV feels incorrect. 

Additionally, shouldn't these controls be grouped, as described in the guidelines right above this section?

Author: > Everything looks so squished horizontally.

I think everything looks so squished because of the text in the diagram. 

---

> Also, the spacing between expiration and CVV feels incorrect.

It definitely is. My mistake.

---

> Additionally, shouldn't these controls be grouped, as described in the guidelines right above this section?

Yup, another mistake on my part.

<Box mb={3} display="flex" alignItems="flex-start" flexDirection={["column", "column", "column", "column", "row"]} sx={{gap: 3}}>
    <img
        width="456"
        alt="Primary discipline radio buttons, selected 'Other' option reveals a text input"
        src="https://user-images.githubusercontent.com/2313998/164065214-0c0bc39f-52e5-4534-9c69-a1f6e4566fa9.png"
    />
    <Box as="p" mt="0">If the parent form control provides sufficient visual context, you may visually hide the label. However, you must specify text for a visually hidden label that is accessible to screen readers.</Box>
</Box>

<Box mb={3} display="flex" alignItems="flex-start" flexDirection={["column", "column", "column", "column", "row"]} sx={{gap: 3}}>
    <div>
        <img
            alt="Checkbox that progressively discloses a dropdown with an internal label"
            src="https://user-images.githubusercontent.com/2313998/164065216-a2702413-5e84-4ec0-b990-b72261df0fb1.png"
        />
        <img
            alt="Checkbox that progressively discloses a labeled text input"
            src="https://user-images.githubusercontent.com/2313998/164065217-7969283a-6dbb-43b6-8f62-e4ab995de246.png"

>         />

Reviewer 2: > <img src="https://user-images.githubusercontent.com/293280/166303568-103e34e7-50a1-4f40-98b0-3c8fa3eeaced.png" alt="" width="400" />

This UI feels confusing, with the checkbox label and text input label having the same alignment and style. I don't think we should use a search icon in that case, but I'm afraid that's not the only thing that's adding to the confusion.

Author: You're right - this isn't a great example. I'll come up with a better way to show a labeled input that is "nested" and progressively disclosed.

    </div>
    <Box as="p" mt="0">If the parent form control does not provide sufficient context, a visible label should be shown. Make sure the label of the nested form control doesn’t clash with the label of its parent form control.</Box>
</Box>

For more information about progressive disclosure, see [Progressive disclosure](/ui-patterns/progressive-disclosure).

## Validation

### Use Primer for validation

> Use Primer instead of browser-native validation UI. Browser-native validation messages are not accessible to screen readers, and they visually clash with Primer styles.

Reviewer 2: Do we need this sub-heading?

It feels confusing to read "Use Primer instead of". Instead, we could refer to something like:

"Primer components have built-in validation support (...)"

Author: Nah, we don't need it. I think I was just trying to emphasize that we **should not** rely on browser validation.

I'll find a better way to do that :)

### Validation statuses

<Box mb={3} display="flex" alignItems="flex-start" flexDirection={["column", "column", "column", "column", "row"]} sx={{gap: 3}}>
  <Box flexGrow={1} flexBasis="0">
    <img
        alt="input with forbidden characters"
        src="https://user-images.githubusercontent.com/2313998/164065220-30c25ca4-827f-4327-95f4-9be888272813.png"
    />
    <Box as="p" mt="0"><strong>Error messages:</strong> An invalid field should always have a message explaining why the value does not pass validation. The message should explain why the value is invalid, and unblock users from completing their task by guiding them to a valid value.</Box>
  </Box>

  <Box flexGrow={1} flexBasis="0">
    <img
        alt="valid repo name success"
        src="https://user-images.githubusercontent.com/2313998/164065221-7e4f6a21-e9a7-465b-9a86-83c61533fdec.png"
    />
    <Box as="p" mt="0"><strong>Success messages:</strong> A success message may be used when a user might need extra assurance that the field's value is valid. For example: when creating a repository name, it’s nice to be assured that the name is available and valid.</Box>
  </Box>
</Box>

### Validation on submit

> The default behavior of the web is to perform validation when the user attempts to submit the form. This lets the user flow quickly through the form without interruption, but the user has to wait until submitting the form to know if they’ve entered invalid data. 

Reviewer 3: ```suggestion
The default behavior of the web is to perform validation when the user attempts to submit the form. This lets the user flow quickly through the form without interruption.
```

When the form fails validation, the first invalid input should be focused and scrolled into the viewport.

> After a form has failed validation, you may switch to inline validation to provide quicker feedback than if you made the user re-submit the form to re-check validation statuses.

Reviewer 3: ```suggestion
After a form has been submitted and failed validation, you may switch to inline validation to provide quicker feedback.
```

### Inline validation

> Inline validation provides immediate feedback as the user is entering data. For example: when setting a username, you could show a success message if the name is available or an error message if it’s taken.

Reviewer 3: ```suggestion
Inline validation provides immediate feedback while the user is entering data. For example: when setting a username, you could show a success message if the name is available or an error message if it’s taken.
```

</Box>

<Box mb={3} display="flex" alignItems="flex-start" flexDirection={["column", "column", "column", "column", "row"]} sx={{gap: 3}}>
    <img
        width="456"
        alt="screen recording of validation being performed as the user tries to fix an invalid input"
        src="https://user-images.githubusercontent.com/2313998/164065860-25e23fa0-e22a-435d-adbd-5666a9fd7890.gif"
    />
    <Box as="p" mt="0">We shouldn’t attempt to validate an input before the user is done with it. So validation may be performed as the user is typing or making their selection, but only <strong>after the first time the input has been validated</strong> and the <strong>input is in an invalid state</strong>. This gives the user early positive feedback by removing the error if the user makes a change that fixes it.</Box>
</Box>

<Box mb={3} display="flex" alignItems="flex-start" flexDirection={["column", "column", "column", "column", "row"]} sx={{gap: 3}}>
    <img
        width="456"
        alt="screen recording of valid repo name input showing loading indicator after blur"
        src="https://user-images.githubusercontent.com/2313998/164065857-573ca6db-f2c7-49ac-a3cf-bf479a40973d.gif"
    />

>     <Box as="p" mt="0">If the form control’s validation is likely to take more than 1 second, show a loading indicator.</Box>

Reviewer 2: > If the form control’s validation is likely to take more than 1 second

Is this the right metric to decide whether to display a loading indicator? What if the process is known as a quick one but the user is on a slow connection?

Reviewer 2: Also, if we're extending the height of the input control based on an async request that may take long to process, how do we avoid the content to reflow unexpectedly?

Author: > Is this the right metric to decide whether to display a loading indicator? What if the process is known as a quick one but the user is on a slow connection?

It shouldn't matter what connection the user is on. This is a detail that would be handled in the implementation.

Author: > if there's a server request, wouldn't it make more sense to show the loading spinner regardless?

@vdepizzol - I was trying to avoid showing too many loading indicators because they could make the app feel slower. If we wait 1 second, we're still giving the user feedback that something is happening, but only if it's something that's making the user wait a noticeable amount of time.

Does that make sense?

</Box>

> <Note variant="warning">

Reviewer 2: I wonder if this note could be simplified a bit, and placed closer to the top of that section before the examples?

>   Inline validation offers immediate feedback, but carefully weigh that with the following drawbacks.

Reviewer 3: ```suggestion
  Inline validation offers immediate feedback, but be sure to consider the following drawbacks.
```

---
title: Forms
---

import {Box, Heading, Link, Text} from '@primer/components'

<Box sx={{fontSize: 3}} class="lead" as="p">
    Forms are used to complete tasks that require data input from the user. For example: creating a new repo, configuring settings, and logging in.
</Box>

<Box sx={{fontSize: 3}} class="lead" as="p">
    Primer's form design guidelines aim to minimize the effort and cognitive load required to complete a task that involves a form.
</Box>

## Form control anatomy

> ![diagrams labeling the anatomy of a text field and a checkbox field](https://user-images.githubusercontent.com/2313998/164065184-e3c615a3-fb68-4878-b7f8-9a57efb748cb.png )

Reviewer 5: I was a bit confused at first looking at this example, because my immediate reaction wasn't thinking that this was a vertical vs. horizontal layout, but rather two different form types (text input vs. radio check). Do we need to classify this as horizontal vs. vertical, or is it enough to describe the different parts of the forms (label, input, ...)? 🤔

Author: Thanks for calling that out. I don't think we need to classify horizontal vs vertical. I see how it's confusing.

<DoDontContainer>
  <Do>
    <img role="presentation" src="https://user-images.githubusercontent.com/2313998/166832268-519ebea3-e989-455c-b182-58e991d78a25.png" />
    <Caption>Hide a caption that is redundant with the validation message</Caption>
  </Do>
  <Dont>
    <img role="presentation" src="https://user-images.githubusercontent.com/2313998/166832269-4756f340-c24a-41bd-8fa9-db807607c874.png" />
    <Caption>Don’t show a repetitive validation message</Caption>
  </Dont>
</DoDontContainer>

For more information about form validation, see the [validation guidelines](#invalid-input).

## Input methods

### Open-ended text

![Text input, open-ended autocomplete text input, plain text input, textarea](https://user-images.githubusercontent.com/2313998/168104196-e0400967-a455-4513-97b0-3005d7ee3d0c.png)

Use an open-ended text field when the field does not have a list of possible values. If the input is able to suggest values, use [autocomplete](#autocomplete) to allow users to pick a value or enter their own.

### A set of selectable options

> ![Action menu, autocomplete input with limited options, checkbox group, radio group, select](https://user-images.githubusercontent.com/2313998/168104199-2990e43c-6204-4ef5-970c-c9babaf1196c.png)

Reviewer 5: As a consumer of this content, `Action menu` and `Select` look pretty close to me, and I have no context to help me select which version I should pick. I wonder if we could provide additional content, or a link to help us make this decision?

Author: This will be covered in the `Action menu` and `Select` component docs.

Show a set of selectable options when there is a finite number of possible values.

## Structure

> Forms should have a structure that makes it easy for users to scan. Form controls that are stacked vertically and left-aligned are easier for sighted users to scan visually.

Reviewer 5: In production, we have a few examples of forms breaking this rule (in `Theme preferences` the radio btns stack horizontally, for example).

The do/don't format reads to me as a _must_ do, whereas it seems like this might be a suggestion instead? 🤔

Reviewer 6: @heyamie I think the goal with interface guidelines is to try and document the ideal pattern even if its not always true today. The theme preference example is a great callout! That code and design will need to be addressed based on this guideline. I wonder if we need some kind of explainer that these do/dont's might not be true in production at the time of documentation 🤔

Author: Great point about horizontal radio buttons. The intention of this "do/don't" is to discourage forms that flow into multiple columns for no reason. I think it's totally fine to have horizontal radio buttons or have two related text inputs like "First name" / "Last name" on the same line.

I'll update the "do/don't" images to be more clear.

    />
    <Box as="p" mt="0">To keep forms concise, you may choose to hide or show form controls based on selections the user has made.</Box>
</Box>

#### Nested form controls

Sometimes, progressively disclosed form controls can be visually nested under a parent form control. A common pattern is to use a checkbox or radio “checked” state to decide whether to show a related form control.

<Box mb={3} display="flex" alignItems="flex-start" flexDirection={["column", "column", "column", "column", "row"]} sx={{gap: 3}}>
    <img
        width="456"
        alt="Primary discipline radio buttons, selected 'Other' option reveals a text input"
        src="https://user-images.githubusercontent.com/2313998/164065214-0c0bc39f-52e5-4534-9c69-a1f6e4566fa9.png"
    />
    <Box as="p" mt="0">If the parent form control provides sufficient visual context, you may visually hide the label. However, you must specify text for a visually hidden label that is accessible to screen readers.</Box>
</Box>

<Box mb={3} display="flex" alignItems="flex-start" flexDirection={["column", "column", "column", "column", "row"]} sx={{gap: 3}}>
    <div>
        <img

>             alt="Checkbox that progressively discloses a dropdown with an internal label"

Reviewer 5: This is the first time I'm being introduced to the `internal label` concept, I wonder if this is something we should highlight above, on the `Labels` section?

Author: This is only something we'd do with an `Action menu` component, so that information will go in that component's guidelines.

    <p>
      An autocomplete input renders a text input that allows a user to quickly filter through a list of options to pick one or more values.
    </p>
      <a href="https://primer.style/react/Autocomplete">Primer React implementation</a>
  </div>
</Box>

<Box mb={4} display="flex" alignItems="flex-start" flexDirection={["column", "column", "column", "column", "row"]} sx={{gap: 3}}>
  <img
    role="presentation"
    width="456"
    src="https://user-images.githubusercontent.com/2313998/166846801-91a03e2d-aedb-4ba3-b8ea-c7abcb7eadd6.png"
  />
  <div>
    <Heading fontSize={3}>Checkbox group</Heading>
    <p>
      A set of checkboxes to let users make one or more selections from a short list of options
    </p>
      <a href="https://primer.style/react/CheckboxGroup">Primer React implementation</a>
  </div>
</Box>

<Box mb={4} display="flex" alignItems="flex-start" flexDirection={["column", "column", "column", "column", "row"]} sx={{gap: 3}}>
  <img
    role="presentation"
    width="456"
    src="https://user-images.githubusercontent.com/2313998/166846802-7d3a0563-d853-4d7b-b26f-0bb4e8c0d3ad.png"
  />
  <div>
    <Heading fontSize={3}>Form control</Heading>
    <p>

>       A form control renders a labelled input and, optionally, associated validation text and/or hint text.

Reviewer 5: This is probably out of scope for this particular task, but I just wanted to raise that if I wanted to have a group of radio inputs, I could pick either `Form control` or `Radio group` 😬 It's not totally clear why I should pick one vs. the other

Author: I will make sure this is clear when we write the guidelines for those components.

Incase you're curious: the main difference is that a `Form control` is used for a single input, and a `Radio group` is used to group multiple related radio inputs. The markup is rendered differently to support assistive technology for screen readers, and the typographic hierarchy is different. Compare this [RadioGroup example](https://primer.style/react/RadioGroup#with-caption) to this [FormControl example](https://primer.style/react/FormControl#with-a-caption).

---
title: Forms
---

import {Box, Heading, Link} from '@primer/components'

<Box sx={{fontSize: 3}} class="lead" as="p">
    Forms are used to complete tasks that require data input from the user. For example: creating a new repo, configuring settings, and logging in.
</Box>

<Box sx={{fontSize: 3}} class="lead" as="p">
    Primer's form design guidelines aim to minimize the effort and cognitive load required to complete a task that involves a form.

> </Box>

Reviewer 2: @mperrotti Can we simplify this leading description and make it into a single paragraph?

> Forms are used to complete tasks that require data input from the user. For example: creating a new repo, configuring settings, and logging in.
>
> Primer's form design guidelines aim to minimize the effort and cognitive load required to complete a task that involves a form.

## Form control anatomy

> ![diagrams labeling the anatomy of a text field and a checkbox field](https://user-images.githubusercontent.com/2313998/168928812-fa5c9ad7-4f92-4e9d-8e6e-62eb4413985c.png)

Reviewer 2: The anatomy image is very blurry on non-high-resolution displays (same for most images on this page in fact). Compare it with the paragraph text in the screenshot:

> ![Screenshot](https://user-images.githubusercontent.com/293280/169879310-48b22366-7858-48aa-8910-bfea1a99cfce.png)

For the ActionList guidelines, I made the screenshots exactly 2x the rendered output. @mperrotti can you make these adjustments?

Author: Good catch. These aren't screenshots, they're rendered and exported from Figma. I'll try bumping the export up to @2x

Reviewer 6: For this particular image, what do you think about pulling those pointer lines out a bit instead of overlapping elements? It looks a little busy to me. You might also consider aligning the label text on one side– so for the lefthand labels (Label, Input) they could all align on the right (end). 

I have an example handy from the Autocomplete docs:

<img width="960" alt="Autocomplete anatomy" src="https://user-images.githubusercontent.com/18661030/170736452-b6d47909-eecf-4f8f-ba5f-481aeff647dc.png">

Reviewer 6: @mperrotti the connector lines between the label and the component with the circle on the end- the way the circle overlaps part of the component (or whatever its pointing to) makes it a little hard to see throughout. Just wondering if it might look more clear to add some white space between the circle and the component like in the screenshot of Autocomplete?

![circle overlapping ui element](https://user-images.githubusercontent.com/18661030/171426141-e7e97c54-0dda-4715-b76a-5f6a572aaf51.png)

Author: Thanks for clearing that up. I think that would be a nice improvement. Will fix.

- Screen readers do not read placeholder text as a label

### Required field indicator

When a field is required to have a value, it should be visibly marked as required. An individual checkbox or radio button cannot be marked as required.

For more information, see the [required fields](#required-fields) section.

### Input (required)

The input is what the user interacts with to set the value of the form control. You may pre-fill inputs with smart default values, but be careful about making too many assumptions about what a user wants or needs.

### Caption

A caption may be used to provide additional context about the field to help users fill in the correct data or explain how the data will be used. Caption text should be as short as possible.

Caption text may be displayed alongside a validation message, or it may be hidden if it only provides redundant information.

Caption text may be used to augment the [label](#label), but should not be redundant with the label or any other parts of the form control. If the caption feels redundant, try removing it. 

<DoDontContainer>
  <Do>

>     <img role="presentation" src="https://user-images.githubusercontent.com/2313998/166832266-c34634ec-2ade-4bf6-8401-df0bdc02b324.png" />

Reviewer 2: ![image](https://user-images.githubusercontent.com/293280/169878988-d9667839-ff30-444a-a88f-e0078bcf42b9.png)

@mperrotti  This "do" example feels quite redundant 🤔

When the form fails validation, guide the user to the invalid inputs:
- If the form has 3 or more errors, you may show an [interactive summary of errors](#interactive-summary-of-errors)
- If an interactive summary of errors is not shown, the first invalid input should be focused and scrolled into the viewport

After a form has been submitted and failed validation, you may switch to inline validation to provide quicker feedback.

#### Interactive summary of errors

<Box mb={3} display="flex" alignItems="flex-start" flexDirection={["column", "column", "column", "column", "row"]} sx={{gap: 3}}>
    <img
        width="456"
        alt="a form for biographical information that has three invalid inputs"
        src="https://user-images.githubusercontent.com/2313998/168926352-c30563ab-3782-4b86-8d21-1ad6dcfa09e3.png"
    />
    <div>
      <Box as="p" mt="0">In a <a href="/ui-patterns/messaging#flash-alerts">flash alert</a> at the top of the form, list the invalid inputs as anchor links. When the link is activated, place focus in it's corresponding input.</Box>
      <Box as="p" mt="0">When the flash alert appears, it should be focused.</Box>
    </div>
</Box>

> ### Inline validation

Author: You're right - all this motion is overwhelming. I'm going to try converting these to `<video>` elements that can be played or paused individually.

---
title: Forms
---

import {Box, Heading, Link, Text} from '@primer/components'

<Box sx={{fontSize: 3}} class="lead" as="p">
    Forms are used to complete tasks that require data input from the user. For example: creating a new repo, configuring settings, and logging in.
</Box>

<Box sx={{fontSize: 3}} class="lead" as="p">
    Primer's form design guidelines aim to minimize the effort and cognitive load required to complete a task that involves a form.
</Box>

## Form control anatomy

![diagrams labeling the anatomy of a text field and a checkbox field](https://user-images.githubusercontent.com/2313998/164065184-e3c615a3-fb68-4878-b7f8-9a57efb748cb.png )

> ### Label (required)

Reviewer 6: In PRC, I see a prop called `layout` which I think is intended to allow the label to display on top of the input, or on the side (left) before the input. Am I correct in that assumption? The code looks a little broken right now:

![image](https://user-images.githubusercontent.com/18661030/167962742-ce31d59a-4246-4f6e-a118-be6a4dec2465.png)

If this is meant to be inline on the left, is that a pattern we need to support? If yes, we might want to document when you would use an inline label.

Author: This prop is meant to be used when you're using a custom input component, and you want the input to be laid out like a checkbox or radio form control. You can see an example in the docs: https://primer.style/react/FormControl#with-a-custom-input

I was stumped by this prop in the component API, as were the other PRC folks I consulted. Do you have any ideas @langermank ? At the very least, we shouldn't show this prop in Storybook stories.

Reviewer 6: Got it, I totally thought this was implying a label inline vs. block.. so its more like a private API. I might have called the prop `labelPosition`? But I haven't thought it through too much 😄 in the meantime, you could always add a description to that prop in Storybook to explain what the does or maybe when to use it.

Caption text may be used to augment the [label](#label), but should not be redundant with the label or any other parts of the form control. If the caption feels redundant, try removing it. 

<DoDontContainer>
  <Do>
    <img role="presentation" src="https://user-images.githubusercontent.com/2313998/166832266-c34634ec-2ade-4bf6-8401-df0bdc02b324.png" />
    <Caption>Use caption text that shows new information and provides helpful context</Caption>
  </Do>
  <Dont>
    <img role="presentation" src="https://user-images.githubusercontent.com/2313998/166832267-e00f88bb-6a81-42e2-a29d-63cc58a70550.png" />
    <Caption>Don’t use caption text that is redundant</Caption>
  </Dont>
</DoDontContainer>

### Validation message

A single validation message may be displayed to provide helpful information for a user to complete their task. For example: explain why a value is invalid so they can correct it and submit the form.

An individual checkbox or radio should not have its own validation.

<DoDontContainer>
  <Do>
    <img role="presentation" src="https://user-images.githubusercontent.com/2313998/164065196-b0146ade-eccb-4a71-b1b6-f4a82b96c115.png" />

>     <Caption>Show a validation message for the group of inputs</Caption>

Reviewer 6: Have we considered a validation pattern where if there are multiple errors, we render a single box above the form with a list of each error- which links to each input with the issue? Its my understanding that this is the preferred method for handling multiple form errors, but perhaps that case hasn't come up for accessibility testing? 

This is from the book [Form Design Patterns](https://www.smashingmagazine.com/printed-books/form-design-patterns/)

![image from Form Design Patterns book showing an error box above a group of fields, linking to each field that is invalid](https://user-images.githubusercontent.com/18661030/167961570-bcc840ee-e4f0-4fb0-adf4-43d383384625.png)

Author: I had considered this pattern, but I tried it on a settings page and it felt like too much error text was being shown at once.

However, this pattern is useful and I think it's worth further exploration. Do you think we could explore it in a separate issue? Or should I do it as part of this PR?

Reviewer 2: @mperrotti have you explored this?

Reviewer 2: >  ![image](https://user-images.githubusercontent.com/293280/169881879-b9e9e92f-696c-4205-8275-e3c428d4f020.png)

@mperrotti additionally, is this "Pick your top two" and "Only two selections are allowed" a good recommendation for checkboxes? It feels out of place — I could see that pattern happening for a larger number ("Pick up to 10 items"?), but two...? Feels like a weird pattern...

Author: I thought I removed this example. If it's still around somewhere, I'll change it.

Author: I see this is still used used for a do/don't example.

<img width="977" alt="Screen Shot 2022-05-23 at 5 59 28 PM" src="https://user-images.githubusercontent.com/2313998/169912361-6a1bce1c-8f72-42a4-adc1-bda234dc168d.png">

  </Do>
  <Dont>
    <img role="presentation" src="https://user-images.githubusercontent.com/2313998/164065199-6cfa4893-55cb-4f81-a7d6-4b6ff66339eb.png" />
    <Caption>Don’t show a validation message for each input</Caption>
  </Dont>
</DoDontContainer>

Information from the caption should not be repeated in the error message. Show the validation message and remove the caption. A validation message makes the field easier to spot when a user is scanning for invalid fields.

<DoDontContainer>
  <Do>
    <img role="presentation" src="https://user-images.githubusercontent.com/2313998/166832268-519ebea3-e989-455c-b182-58e991d78a25.png" />
    <Caption>Hide a caption that is redundant with the validation message</Caption>
  </Do>
  <Dont>
    <img role="presentation" src="https://user-images.githubusercontent.com/2313998/166832269-4756f340-c24a-41bd-8fa9-db807607c874.png" />
    <Caption>Don’t show a repetitive validation message</Caption>
  </Dont>
</DoDontContainer>

For more information about form validation, see the [validation guidelines](#invalid-input).

> ## Input methods

Reviewer 6: Before this is published, it would probably be good to update the focus styles to the new design (this should already exist in Figma today)

Author: Great point. We really need to come up with a more future-proof way of rendering pattern examples.

cc @tallys - this is a relevant to the documentation image strategies I've been mentioning lately.

</Box>

<Box mb={3} display="flex" alignItems="flex-start" flexDirection={["column", "column", "column", "column", "row"]} sx={{gap: 3}}>
    <img
        width="456"
        alt="screen recording of valid repo name input showing loading indicator after blur"
        src="https://user-images.githubusercontent.com/2313998/164065857-573ca6db-f2c7-49ac-a3cf-bf479a40973d.gif"
    />
    <Box as="p" mt="0">If the form control’s validation is likely to take more than 1 second, show a loading indicator.</Box>
</Box>

### Validating a group of form controls

<Box mb={3} display="flex" alignItems="flex-start" flexDirection={["column", "column", "column", "column", "row"]} sx={{gap: 3}}>
    <img
        width="456"
        alt="checkbox group with an error"
        src="https://user-images.githubusercontent.com/2313998/164065224-d1ba56ce-c632-4318-b9b2-b75940f771a6.png"
    />

>     <Box as="p" mt="0">A validation message may be used to describe an error for a group of form controls. For example: if a user checks 3 checkboxes when they’re only allowed to check 2 options.</Box>

Reviewer 6: I think my comment above about grouping error messages could be described in this section instead. The case you mention here is a single error message for a group of controls, but we should document how to present multiple individual error messages for a group of controls as well.

Reviewer 6: Here's a screenshot from a previous form I worked on:

![Contact form with multiple fields showing an error state. At the top of the form, there's a group with a title and list of links to each field, stating what the error is with the field](https://user-images.githubusercontent.com/18661030/168659393-55b9a577-3dee-44e8-b2d1-bf03ff326cfb.png)

I think the idea here is if there are multiple errors in a single `<form>`, show them in a labeled group with links to each field. My original comment might have sounded like multiple errors per _field_, but that's not what I mean 😅 that feels problematic in a different way!

@alliethu any thoughts here? I pulled this from a book on accessible forms, and that was a year ago. I haven't researched this recently.

Reviewer: @langermank Your annotation is very much inline with what I would have done myself. I'm attaching a rough annotation below that includes a few more details: 

![Error summary notation](https://user-images.githubusercontent.com/7228610/168684134-1c58188a-e8af-4349-bcbc-cf04cbeedb1a.png)

Considerations to make:
- add `role="alert"` to the summary `div`
- move keyboard focus to the error summary
- use a semantic heading (I'm thinking an `h2` to title the error summary area)
- reference the heading in the summary `div` by using `aria-labelledby`
- use a semantic list for the error summaries
- when a link is activated, place focus in its corresponding input field
- the error messages should be worded the same in the error summary and inline with each input.
cc/ @mperrotti

Author: Thank you both for the helpful examples! I misunderstood what was being suggested. I thought @langermank was suggesting this pattern for a single group of checkbox or radio inputs.

It feels aggressive to _always_ show the error summary alert banner. We should explain when a summary is needed and when it isn't. Maybe when there are 3 or more errors?

Author: I'm noticing repetitive error messages in @alliethu's example, and that feels like it could be overwhelming or confusing to sighted and non-sighted users.

Author: I pushed [a commit](https://github.com/primer/design/pull/261/commits/0c55099ae9f157dbdc2c9857d1e1070638ecbdf6) to add some guidance on designing error summaries. I'm not feeling great about it yet, and I'd appreciate some time pairing with somebody from the accessibility team.

</Box>

## Submission

Forms should follow consistent patterns for [submitting and saving data](/ui-patterns/saving).

## Related components

<Box mb={4} display="flex" alignItems="flex-start" flexDirection={["column", "column", "column", "column", "row"]} sx={{gap: 3}}>

>   <img

Reviewer 6: I'm not sure an image of ActionMenu without any selection makes sense here- I think if we want to show ActionMenu with selection, the image could include one of those states (multi or single select)

    />
    <Caption>Don’t lay out forms that flow into columns just to reduce the vertical space used by the form</Caption>
  </Dont>
</DoDontContainer>

### Order of form controls

Form fields should be in a predictable order that flows intuitively.

To achieve an intuitive flow:

- Order fields by their relative level of importance
- Keep related fields near each other
- When possible, keep inputs that require keyboard input near each other so that users who use a mouse don't have to switch back and forth between clicking and typing

### Sizing form controls

#### Hint at the expected length of the value

<Box
  mb={3}
  display="flex"
  alignItems="flex-start"
  flexDirection={['column', 'column', 'column', 'column', 'row']}
  sx={{gap: 3}}
>
  <img
    width="456"
    role="presentation"
    src="https://user-images.githubusercontent.com/2313998/170069941-0e7ead2f-1e20-47dc-b7b0-e7a27480d2a5.png"
  />
  <Box>
    <Box as="p" mt="0">
      Form controls should be sized to fit their value. Start with the browser default width, and adjust as needed.
    </Box>
    <Box as="ul" pl="1rem">
      <li>

>         Single-line text inputs get a default width set by the browser, but their width can be changed to fit the

Reviewer 6: What is the default width? It might be helpful to mention that here.

Author: We don't define it - It's set by the browser.
